File: 482.py
class Solution:
    def licenseKeyFormatting(self, s: str, k: int) -> str:
        # time: O(n), space: O(1)
        s = s.upper().replace("-", "")[::-1]
        groups = []
        cur = ""
        for c in s:
            cur += c
            if len(cur) == k:
                groups.append(cur[::-1])
                cur = ""
        if cur:
            groups.append(cur[::-1])
        return "-".join(reversed(groups))

File: 482.java
class Solution {
    public String licenseKeyFormatting(String s, int k) {
        // time: O(n), space: O(1)
        int cnt = 0;
        int n = s.length();
        StringBuilder res = new StringBuilder();
        for (int i = n - 1; i >= 0; i--) {
            if (s.charAt(i) != '-') {
                res.append(Character.toUpperCase(s.charAt(i)));
                cnt++;
                if (cnt == k) {
                    res.append('-');
                    cnt = 0;
                }
            }
        }
        if (res.length() > 0 && res.charAt(res.length() - 1) == '-') {
            res = new StringBuilder(res.substring(0, res.length() - 1));
        }
        res.reverse();
        return res.toString();
    }
}